{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook serves to test functions for connecting to and retrieving information from [deepBlue](http://deepblue.mpi-inf.mpg.de/). The following functionality needs to be supported:\n",
    "\n",
    " * Connect to the server and request a given sample.\n",
    " * Retrieve chromosome names/sizes\n",
    " * Retrive the signal of a region, returning a numpy array\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    # python 2\n",
    "    import xmlrpclib\n",
    "except:\n",
    "    # python 3\n",
    "    import xmlrpc.client as xmlrpclib\n",
    "import time\n",
    "import numpy as np\n",
    "\n",
    "class deepBlue(object):\n",
    "    def __init__(self, sample, url=\"http://deepblue.mpi-inf.mpg.de/xmlrpc\", userKey=\"anonymous_key\"):\n",
    "        \"\"\"\n",
    "        Connect to the requested deepblue server with the given user key and request the specifed sample from it.\n",
    "        \"\"\"\n",
    "        self.sample = sample\n",
    "        self.url = url\n",
    "        self.userKey = userKey\n",
    "        self.server = xmlrpclib.Server(url, allow_none=True)\n",
    "        self.info = None\n",
    "        self.experimentID = None\n",
    "        self.genome = None\n",
    "        self.chromsDict = None\n",
    "\n",
    "        # Set self.experimentID\n",
    "        experimentID = self.getEID()\n",
    "        if not experimentID:\n",
    "            raise RuntimeError(\"The requested sample({}) has no associated experiment!\".format(sample))\n",
    "\n",
    "        # Set self.info\n",
    "        (status, resp) = self.server.info(self.experimentID, userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received the following error while fetching information about '{}': {}\".format(resp, sample))\n",
    "        self.info = resp[0]\n",
    "\n",
    "        # Set self.genome\n",
    "        genome = self.getGenome()\n",
    "        if not genome:\n",
    "            raise RuntimeError(\"Unable to determine an appropriate genome for '{}'\".format(sample))\n",
    "\n",
    "        # Set self.chroms\n",
    "        chroms = self.getChroms()\n",
    "        if not chroms:\n",
    "            raise RuntimeError(\"Unable to determine chromosome names/sizes for '{}'\".format(sample))\n",
    "\n",
    "    def getEID(self):\n",
    "        \"\"\"\n",
    "        Given a sample name, return its associated experiment ID (or None on error).\n",
    "\n",
    "        self.experimentID is then the internal ID (e.g., e52525)\n",
    "        \"\"\"\n",
    "        (status, resps) = self.server.search(self.sample, \"experiments\", self.userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received an error ({}) while searching for the experiment associated with '{}'\".format(resps, self.sample))\n",
    "        for resp in resps:\n",
    "            if resp[1] == self.sample:\n",
    "                self.experimentID = resp[0]\n",
    "                return resp[0]\n",
    "        return None\n",
    "\n",
    "    def getGenome(self):\n",
    "        \"\"\"\n",
    "        Determines and sets the genome assigned to a given sample. On error, this raises a runtime exception.\n",
    "\n",
    "        self.genome is then the internal genome ID.\n",
    "        \"\"\"\n",
    "        if \"genome\" in self.info.keys():\n",
    "            self.genome = self.info[\"genome\"]\n",
    "        return self.genome\n",
    "\n",
    "    def getChroms(self):\n",
    "        \"\"\"\n",
    "        Determines and sets the chromosome names/sizes for a given sample. On error, this raises a runtime exception.\n",
    "\n",
    "        self.chroms is then a dictionary of chromosome:length pairs\n",
    "        \"\"\"\n",
    "        (status, resp) = self.server.chromosomes(self.genome, self.userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received an error while fetching chromosome information for '{}': {}\".format(self.sample, resp))\n",
    "        self.chromsDict = {k: v for k, v in resp}\n",
    "        return resp\n",
    "\n",
    "    def chroms(self, chrom=None):\n",
    "        \"\"\"\n",
    "        Like the chroms() function in pyBigWig, returns either chromsDict (chrom is None) or the length of a given chromosome\n",
    "        \"\"\"\n",
    "        if chrom is None:\n",
    "            return self.chromsDict\n",
    "        elif chrom in self.chromsDict:\n",
    "            return self.chromsDict[chrom]\n",
    "        return None\n",
    "\n",
    "    def values(self, chrom, start=0, end=0):\n",
    "        \"\"\"\n",
    "        Return a numpy array with values for each base in the requested region.\n",
    "        If end==0 then the region from start to the end of the chromosome is returned.\n",
    "        \"\"\"\n",
    "        if chrom not in self.chroms():\n",
    "            raise RuntimeError(\"'{}' is not a valid chromosome.\".format(chrom))\n",
    "\n",
    "        if end == 0:\n",
    "            end = self.chromsDict[chrom]\n",
    "\n",
    "        if start >= end:\n",
    "            raise RuntimeError(\"The start position MUST be less then the end position ({} and {})\".format(start, end))\n",
    "\n",
    "        intervals = self.intervals(chrom, start, end)\n",
    "\n",
    "        # Generate the output\n",
    "        o = np.empty(end - start)\n",
    "        o[:] = np.nan\n",
    "        for interval in intervals:\n",
    "            s = interval[0] - start\n",
    "            e = s + interval[1] - interval[0]\n",
    "            if s < 0:\n",
    "                s = 0\n",
    "            if e >= end - start:\n",
    "                e = end - start\n",
    "            if e > s:\n",
    "                o[s:e] = interval[2]\n",
    "\n",
    "        return o\n",
    "\n",
    "    def intervals(self, chrom, start=0, end=0):\n",
    "        \"\"\"\n",
    "        Return all intervals overlapping a given region\n",
    "        \"\"\"\n",
    "        if chrom not in self.chroms():\n",
    "            raise RuntimeError(\"'{}' is not a valid chromosome.\".format(chrom))\n",
    "\n",
    "        if end == 0:\n",
    "            end = self.chromsDict[chrom]\n",
    "\n",
    "        if start >= end:\n",
    "            raise RuntimeError(\"The start position MUST be less then the end position ({} and {})\".format(start, end))\n",
    "\n",
    "        # Get the experiment information\n",
    "        (status, queryID) = self.server.select_experiments(self.sample, chrom, None, None, self.userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received the following error while fetching values in the range {}:{}-{} in file '{}': {}\".format(chrom, start, end, self.sample, queryID))\n",
    "        if not queryID:\n",
    "            raise RuntimeError(\"Somehow, we received None as a query ID (range {}:{}-{} in file '{}')\".format(chrom, start, end, self.sample))\n",
    "\n",
    "        # Filter by overlap\n",
    "        (status, filterID) = self.server.input_regions(self.genome, \"{}\\t{}\\t{}\".format(chrom, start, end), self.userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received the following error while setting the region filter {}:{}-{} in file '{}': {}\".format(chrom, start, end, self.sample, filterID))\n",
    "        (status, intersectID) = self.server.intersection(queryID, filterID, self.userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received the following error while setting performing the intersect on file '{}': {}\".format(self.sample, intersectID))\n",
    "\n",
    "        # Query the regions\n",
    "        (status, reqID) = self.server.get_regions(intersectID, \"START,END,VALUE\", self.userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received the following error while fetching regions in the range {}:{}-{} in file '{}': {}\".format(chrom, start, end, self.sample, reqID))\n",
    "\n",
    "        # Wait for the server to process the data\n",
    "        (status, info) = self.server.info(reqID, self.userKey)\n",
    "        request_status = info[0][\"state\"]\n",
    "        while request_status != \"done\" and request_status != \"failed\":\n",
    "            time.sleep(1)\n",
    "            (status, info) = self.server.info(reqID, self.userKey)\n",
    "            request_status = info[0][\"state\"]\n",
    "\n",
    "        # Get the actual data\n",
    "        (status, resp) = self.server.get_request_data(reqID, self.userKey)\n",
    "        if status != \"okay\":\n",
    "            raise RuntimeError(\"Received the following error while fetching data in the range {}:{}-{} in file '{}': {}\".format(chrom, start, end, self.sample, resp))\n",
    "\n",
    "        o = []\n",
    "        for intervals in resp.split(\"\\n\"):\n",
    "            interval = intervals.split(\"\\t\")\n",
    "            if interval[0] == '':\n",
    "                continue\n",
    "            o.append((int(interval[0]), int(interval[1]), float(interval[2])))\n",
    "        return o\n",
    "\n",
    "    def stats(self, chrom, start=0, end=0):\n",
    "        \"\"\"\n",
    "        Like stats() from pyBigWig, but only ever returns the mean\n",
    "        \"\"\"\n",
    "        vals = self.values(chrom, start, end)\n",
    "        if np.all(np.isnan(vals)):\n",
    "            return None\n",
    "        rv = np.nanmean(vals)\n",
    "        if np.isnan(rv):\n",
    "            return None\n",
    "        return [rv]\n",
    "\n",
    "    def close(self):\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(3050021, 3050029, 1.3), (3050029, 3050035, 1.2), (3050035, 3050043, 1.1), (3050043, 3050080, 1.0), (3050080, 3050107, 1.1)]\n"
     ]
    }
   ],
   "source": [
    "#sample = \"GC_T14_10.CPG_methylation_calls.bs_call.GRCh38.20150707.wig\"\n",
    "sample = \"S002R5H1.ERX300721.H3K4me3.bwa.GRCh38.20150528.bedgraph\"\n",
    "#sample = \"S002R5H1.ERX337057.Input.bwa.GRCh38.20150528.bedgraph\"\n",
    "db = deepBlue(sample)\n",
    "vals = db.intervals(\"chr1\", 3050022, 3050100)\n",
    "print(vals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True]\n"
     ]
    }
   ],
   "source": [
    "vals = db.values(\"chr1\", 3050022, 3050100)\n",
    "correct = np.concatenate([[1.3] * 7, [1.2] * 6, [1.1] * 8, [1.0] * 37, [1.1] * 20])\n",
    "print(vals == correct)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
